File: src/boj/boj2098/Main.kt
package boj.boj2098

import java.io.*
import java.util.*
import kotlin.math.min

val MAX_VALUE = 987654321

fun main(args: Array<String>) {
    val br = BufferedReader(InputStreamReader(System.`in`))

    val N = br.readLine().toInt()
    var st: StringTokenizer
    val dist = Array(N, {_ -> st = StringTokenizer(br.readLine()); Array(N, {_ -> st.nextToken().toInt()})})
    print(solve(N, dist))
    br.close()
}

fun TSP(cur: Int, visited: Int, cache: Array<Array<Int>>, dist: Array<Array<Int>>, N: Int): Int {
    if (visited == ((1 shl N) - 1)) {
        return dist[cur][0]
    }
    if (cache[cur][visited] != 0) {
        return cache[cur][visited]
    }
    var ret = MAX_VALUE
    for (next in 0 until N) {
        if ((visited and (1 shl next)) > 0) {
            continue
        }
        if (dist[cur][next] == 0) {
            continue
        }
        ret = min(ret, TSP(next, visited or (1 shl next), cache, dist, N) + dist[cur][next] )

    }
    cache[cur][visited] = ret
    return ret
}

fun solve(N: Int, dist: Array<Array<Int>>): Int {
    val cache = Array(N, {_ -> Array(1 shl N, {_ -> 0})})
    return TSP(0, 1, cache, dist, N)

}

File: src/boj/boj2098/Main.py
import sys
inf = 10**20

def LI(): return [int(x) for x in sys.stdin.readline().split()]
def I(): return int(sys.stdin.readline())

def TPS(cur, visited, cache, dist, N):
    if visited == (1<<N)-1:
        return dist[cur][0]
    if cache[cur][visited] != 0:
        return cache[cur][visited]
    ret = inf
    for next in range(N):
        if visited & 1 << next:
            continue
        if dist[cur][next] == 0:
            continue
        ret = min(ret, TPS(next, visited | (1 << next), cache, dist, N) + dist[cur][next])
    cache[cur][visited] = ret
    return ret

def solve(N, dist):
    cache = [[0 for _ in range(1 << N)] for _ in range(N)]
    return TPS(0, 1, cache, dist, N)

def main():
    N = I()
    dist = [LI() for _ in range(N)]
    print(solve(N, dist))

main()
